### PR TITLE
[InstCombine] Fix a crash in `PointerReplacer`

### PR DESCRIPTION
**Pull request by shiltian**
A crash could happen in `PointerReplacer::replace` when constructing a new
select instruction and there is no replacement for one of its operand. This can
happen when the operand is a load instruction that has been replaced earlier
such that the operand itself is already the new value. In this case, it is not
in the replacement map and `getReplacement` simply returns nullptr.

Fix SWDEV-472192.

### CODE REVIEW

**Comment by shiltian** [comment]
* **#98987** <a href="https://app.graphite.dev/github/pr/llvm/llvm-project/98987?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @shiltian and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: llvm/test/Transforms/InstCombine/select-from-load.ll [hunk 1]
```diff
@@ -0,0 +1,33 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py
+; RUN: opt -passes=instcombine -S -o - %s | FileCheck %s
+; REQUIRES: amdgpu-registered-target
```
Comment (arsenm) [inline llvm/test/Transforms/InstCombine/select-from-load.ll:3]: Move to AMDGPu subdirectory instead

**Review by arsenm — COMMENTED** [review]
Is this the same as https://github.com/llvm/llvm-project/pull/98456

File: llvm/test/Transforms/InstCombine/select-from-load.ll [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+; CHECK-NEXT:    br label [[FOR_COND10:%.*]]
+; CHECK:       for.cond10:
+; CHECK-NEXT:    [[TMP2:%.*]] = load i64, ptr [[TMP1]], align 8
+; CHECK-NEXT:    store i64 [[TMP2]], ptr addrspace(1) null, align 8
+; CHECK-NEXT:    br label [[FOR_COND10]]
+;
+entry:
+  %coerce = alloca %anon, addrspace(5)
+  call void @llvm.memcpy.p5.p4.i64(ptr addrspace(5) %coerce, ptr addrspace(4) %0, i64 0, i1 false)
+  %1 = addrspacecast ptr addrspace(5) %coerce to ptr
```
Comment (arsenm) [inline llvm/test/Transforms/InstCombine/select-from-load.ll:22]: Use named values in tests

File: llvm/test/Transforms/InstCombine/select-from-load.ll (hunk 1 above)
Comment (arsenm) [reply llvm/test/Transforms/InstCombine/select-from-load.ll:3]: Although maybe this doesn't require the target, maybe just the datalayout with A5?

File: llvm/test/Transforms/InstCombine/select-from-load.ll [hunk 3]
```diff
@@ -0,0 +1,33 @@
+; NOTE: Assertions have been autogenerated by utils/update_test_checks.py
+; RUN: opt -passes=instcombine -S -o - %s | FileCheck %s
+; REQUIRES: amdgpu-registered-target
+
+target triple = "amdgcn-amd-amdhsa"
+
+%anon = type { i32, [8 x ptr], ptr }
+
+define amdgpu_kernel void @foo(ptr addrspace(4) byref(%anon) align 8 %0) {
```
Comment (arsenm) [inline llvm/test/Transforms/InstCombine/select-from-load.ll:9]: Probably don't need the CC

**Comment by jofrn** [comment]
> Is this the same as #98456

It is. @shiltian's change is much simpler.

File: llvm/test/Transforms/InstCombine/select-from-load.ll (hunk 1 above)
Comment (shiltian) [reply llvm/test/Transforms/InstCombine/select-from-load.ll:3]: Well, it still needs target. I tried only with data layout and it doesn't reproduce the crash.

**Review by gandhi56 — APPROVED** [review]

**Pull request merged**